test/testframe.c

```diff
@@ -506,326 +569,405 @@ TestParseCmdLine(int argc, char *argv[])
+            ret_value = FAIL;
+            goto done;
         }
     }
-    
-    UpdateTestStats(thread_args);
-    
-    /* Clean up */
-    for (int i = 0; i < GetTestMaxNumThreads(); i++) {
-        free(thread_args[i].test_outcomes);
-        thread_args[i].test_outcomes = NULL;
-    
-        free(thread_args[i].test_descriptions);    
-        thread_args[i].test_descriptions = NULL;
-    }
 
-    free(threads);
-    free(thread_args);
-    
-    threads = NULL;
-    thread_args = NULL;
+    /* Aggregate test results from threads */
+    for (int thread_idx = 0; thread_idx < num_threads; thread_idx++) {
+        struct ThreadPrivData_t *priv_data;
+        size_t                   thread_err_cnt;
 
-    return;
-}
+        priv_data      = thread_args[thread_idx].ThreadLocalParams->MtTestParams.ThreadPrivData;
+        thread_err_cnt = thread_args[thread_idx].ThreadLocalParams->MtTestParams.ThreadErrCnt;
 
-static void
-UpdateTestStats(TestThreadArgs *thread_args) {
-    test_outcome_t final_results[H5_MAX_NUM_SUBTESTS];
-    memset(final_results, (int) TEST_UNINIT, H5_MAX_NUM_SUBTESTS * sizeof(test_outcome_t));
+        if (thread_args[thread_idx].TestRet < 0 || thread_err_cnt > 0)
+            thread_fail = true;
 
-    /* If test does not publish its results information to a threadlocal variable,
-     * do not track statistics */
-    if (thread_args[0].num_tests == 0) {
-        return;
+        if (thread_args[thread_idx].TestRet != SKIP)
+            all_skip = false;
+
+        if (thread_args[thread_idx].TestRet != SUCCEED &&
+            thread_args[thread_idx].TestRet != FAIL &&
+            thread_args[thread_idx].TestRet != SKIP)
+            MESSAGE(2, ("** invalid return value (%d) from thread %d for test %s (%s) \n",
+                thread_args[thread_idx].TestRet, thread_idx, threaded_test->Name,
+                threaded_test->Description));
+
+        if (thread_idx == 0)
+            min_subtests = priv_data->subtest_count;
+        else
+            min_subtests = MIN(min_subtests, priv_data->subtest_count);
     }
 
-    /* Verify that each thread reported the same number of subtests */
-    for (int i = 0; i < GetTestMaxNumThreads(); i++) {
-        if (thread_args[i].num_tests != thread_args[0].num_tests) {
-            fprintf(stderr, "Thread %d reported %ld subtests, but thread 0 reported %ld\n", i, thread_args[i].num_tests, thread_args[0].num_tests);
-            exit(EXIT_FAILURE);
+    /* Ensure value is within range */
+    min_subtests = MIN(min_subtests, TESTFRAME_MAX_NUM_SUBTESTS);
+
+    /* Print delayed sub-test headers, if any, up to the minimum number ran among all threads */
+    if (!all_skip && min_subtests > 0) {
+        for (size_t subtest_idx = 0; subtest_idx < min_subtests; subtest_idx++) {
+            struct ThreadPrivData_t *priv_data;
+
+            priv_data = thread_args[0].ThreadLocalParams->MtTestParams.ThreadPrivData;
+            if (priv_data->subtest_descriptions[subtest_idx])
```

Comment:
This logic is essentially what @mattjala had before, just moved into a field in a thread's copy of the `TestParams_t` structure rather than in actual thread-local data. The logic also doesn't print out `PASSED` / `FAILED` / `-SKIP-` for each of the sub-tests since the formatting can get messy.